ForwardingDecision.cs

```diff
@@ -0,0 +1,43 @@
+/*
+Pax : tool support for prototyping packet processors
+Nik Sultana, Cambridge University Computer Lab, June 2016
+
+Use of this source code is governed by the Apache 2.0 license; see LICENSE.
+*/
+
+namespace Pax {
+
+  // A packet processor makes some sort of forwarding decision, in addition
+  // to possibly analysing/modifying the packet (and generating new packets).
+  // The forwarding decision makes clear whether the processor recommends to
+  // * drop the packet (i.e., nothing more is to be done with it)
+  // * forward to a single port
+  // * forward to multiple ports
+  //
+  // NOTE this idiom simulates using algebraic types in C#. It's values are
+  //      analysed using the Visitor pattern.
+  public abstract class ForwardingDecision {
+    private ForwardingDecision() {}
+
+    // FIXME can use singleton to avoid multiple allocations of this?
```

Comment:
I'm in two minds about mixing different patterns together (visitor and 
singleton in this case). I wonder if there are clean ways to do this. 
Perhaps ultimately the solution will be ad-hoc to this type.

On 2016-07-12 14:10, TMVector wrote:
> In ForwardingDecision.cs [1]:
> 
>> +
>> +namespace Pax {
>> +
>> + // A packet processor makes some sort of forwarding decision, in
> addition
>> + // to possibly analysing/modifying the packet (and generating new
> packets).
>> + // The forwarding decision makes clear whether the processor
> recommends to
>> + // * drop the packet (i.e., nothing more is to be done with it)
>> + // * forward to a single port
>> + // * forward to multiple ports
>> + //
>> + // NOTE this idiom simulates using algebraic types in C#. It's
> values are
>> + // analysed using the Visitor pattern.
>> + public abstract class ForwardingDecision {
>> + private ForwardingDecision() {}
>> +
>> + // FIXME can use singleton to avoid multiple allocations of this?
> 
> I think this FIXME could be done in this PR.
> 
> --
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [2], or mute the
> thread [3].
> 
> 	*
> 
> Links:
> ------
> [1] https://github.com/niksu/pax/pull/7#discussion_r70434050
> [2]
> https://github.com/niksu/pax/pull/7/files/3d69508e501b5f605e634c939d226987d76f81e6#r70434050
> [3]
> https://github.com/notifications/unsubscribe/ABDTTjyWjRk_aGiZF70jQmK4aB5iNS4uks5qU5JIgaJpZM4JKIVi

-- 
http://www.cl.cam.ac.uk/~ns441/